### PR TITLE
Remove gas price

### PR DESCRIPTION
**Pull request by gterzian**
FIX https://github.com/massalabs/massa/issues/3140

* [ ] document all added functions
* [ ] try in sandbox /simulation/labnet
* [ ] unit tests on the added/changed features
  * [ ] make tests compile
  * [ ] make tests pass 
* [ ] add logs allowing easy debugging in case the changes caused problems
* [ ] if the API has changed, update the API specification

### CODE REVIEW

**Comment by gterzian** [comment]
@massalabs/core-team @damip Still have to fix a few tests that are failing(see those currently ignored), sorry I still cannot access discord...

File: massa-async-pool/src/lib.rs [hunk 1]
```diff
@@ -16,33 +16,32 @@
 //!
 //! ```json
 //! {
-//!     "sender": "xxxx",  // address that sent the message and spent max_gas*gas_price+coins on emission
+//!     "sender": "xxxx",  // address that sent the message and spent max_gas+coins on emission
```
Comment (damip) [inline massa-async-pool/src/lib.rs:19]: `max_gas` is not an amount of coins, but a quantity of gas.

=> address that sent the message and spent `fee+coins` on emission

File: massa-async-pool/src/lib.rs [hunk 2]
```diff
@@ -16,33 +16,32 @@
...
 //!     "slot": {"period": 123455, "thread": 11},  // slot at which the message was emitted
 //!     "emission_index": 212,  // index of the message emitted in this slot
 //!     "destination": "xxxx",  // target address
 //!     "handler": "handle_message",  // name of the handler function to call in the target SC
 //!     "validity_start": {"period": 123456, "thread": 12},  // the message can be handled starting from the validity_start slot (included)
 //!     "validity_end": {"period": 123457, "thread": 16},  // the message can be handled until the validity_end slot (excluded)
 //!     "max_gas": 12334,  // max gas available when the handler is called
-//!     "gas_price": "124.23",  // gas price for the handler call
```
Comment (damip) [inline massa-async-pool/src/lib.rs:27]: this needs to be replaced by a `fee: Amount` property

File: massa-async-pool/src/lib.rs [hunk 3]
```diff
@@ -16,33 +16,32 @@
...
 //!     "coins": "1111.11",  // amount of coins to transfer to the destination address when calling its handler
 //!     "data": { ... any object ... }  // data payload of the message, passed as the sole parameter of the destination handler when called
 //! }
 //! ```
 //!
 //! ## How to send a message during bytecode execution
 //!
-//! * messages are sent using an ABI: `send_message(target_address, target_handler, validity_start, validity_end, max_gas, gas_price, coins, data: JSON string) -> Result<(), ABIReturnError>`. Note that data has a configuration defined `max_async_message_data_size`.
+//! * messages are sent using an ABI: `send_message(target_address, target_handler, validity_start, validity_end, max_gas, coins, data: JSON string) -> Result<(), ABIReturnError>`. Note that data has a configuration defined `max_async_message_data_size`.
```
Comment (damip) [inline massa-async-pool/src/lib.rs:34]: `gas_price` needs to be replaced by a `fee` property

File: massa-async-pool/src/lib.rs [hunk 4]
```diff
@@ -16,33 +16,32 @@
...
 //! * when called, this ABI does this:
 //!   * it consumes `compute_gas_cost_of_message_storage(context.current_slot, validity_end_slot)` of gas in the current execution. This allows making the message emission more gas-consuming when it requires storing the message in queue for longer
-//!   * it consumes `max_gas * gas_price + coins` coins from the sender
+//!   * it consumes `max_gas + coins` coins from the sender
```
Comment (damip) [inline massa-async-pool/src/lib.rs:37]: it consumes `fee + coins`

File: massa-async-pool/src/lib.rs [hunk 5]
```diff
@@ -16,33 +16,32 @@
...
 //!   * it generates an `AsyncMessage` and stores it in an asynchronous pool
 //!
-//! Note that `max_gas*gas_price` coins are burned when sending the message.
+//! Note that `max_gas` coins are burned when sending the message.
```
Comment (damip) [inline massa-async-pool/src/lib.rs:40]: Gas and coins are fundamentally different resources.

`fee` coins are burned when sending the message

File: massa-async-pool/src/lib.rs [hunk 6]
```diff
@@ -16,33 +16,32 @@
...
 //!
 //! ## How is the `AsyncPool` handled
 //! ```md
-//! * In the AsyncPool, Messages are kept sorted by `priority = AsyncMessageId(msg.max_gas * msg.gas_price, rev(msg.slot), rev(msg.emission_index))`
+//! * In the AsyncPool, Messages are kept sorted by `priority = AsyncMessageId(msg.max_gas, rev(msg.slot), rev(msg.emission_index))`
```
Comment (damip) [inline massa-async-pool/src/lib.rs:44]: The main sorting criterion should be `Ratio(fee.to_raw(), max(max_gas, 1u64))`.

File: massa-async-pool/src/lib.rs [hunk 7]
```diff
@@ -55,7 +54,6 @@
 //!     * credit target_address with M.coins
 //!     * run the target handler function with M.payload as parameter and the context:
 //!       * max_gas = M.max_gas
-//!       * gas_price = M.gas_price
```
Comment (damip) [inline massa-async-pool/src/lib.rs:58]: fee = M.fee

File: massa-async-pool/src/message.rs [hunk 8]
```diff
@@ -23,19 +23,19 @@ use std::ops::Bound::{Excluded, Included};
 
 /// Unique identifier of a message.
 /// Also has the property of ordering by priority (highest first) following the triplet:
-/// `(rev(max_gas*gas_price), emission_slot, emission_index)`
-pub type AsyncMessageId = (std::cmp::Reverse<Amount>, Slot, u64);
+/// `(rev(max_gas), emission_slot, emission_index)`
```
Comment (damip) [inline massa-async-pool/src/message.rs:26]: change it to the formula with the ratio mentioned in the comment above

File: massa-async-pool/src/message.rs [hunk 9]
```diff
@@ -23,19 +23,19 @@ use std::ops::Bound::{Excluded, Included};
...
+pub type AsyncMessageId = (std::cmp::Reverse<u64>, Slot, u64);
```
Comment (damip) [inline massa-async-pool/src/message.rs:27]: change it to the formula with the ratio mentioned in the comment above

File: massa-async-pool/src/message.rs [hunk 10]
```diff
@@ -23,19 +23,19 @@ use std::ops::Bound::{Excluded, Included};
...
     u64_serializer: U64VarIntSerializer,
 }
 
 impl AsyncMessageIdSerializer {
     pub fn new() -> Self {
         Self {
-            amount_serializer: AmountSerializer::new(),
+            max_gas_serializer: U64VarIntSerializer::new(),
             slot_serializer: SlotSerializer::new(),
             u64_serializer: U64VarIntSerializer::new(),
```
Comment (damip) [inline massa-async-pool/src/message.rs:40]: emission_index_serializer

Comment (damip) [inline massa-async-pool/src/message.rs:40]: let's rename this one emission_index_serializer

File: massa-client/src/cmds.rs [hunk 11]
```diff
@@ -824,19 +823,18 @@ impl Command {
...
                 let max_gas = parameters[2].parse::<u64>()?;
-                let gas_price = parameters[3].parse::<Amount>()?;
-                let fee = parameters[4].parse::<Amount>()?;
+                let fee = parameters[3].parse::<Amount>()?;
 
                 if !json {
-                    match gas_price
-                        .checked_mul_u64(max_gas)
-                        .and_then(|x| x.checked_add(fee))
+                    match Amount::from_str(&max_gas.to_string())
```
Comment (damip) [inline massa-client/src/cmds.rs:835]: Gas is not an amount of coins. This check should be done on `fee` property

File: massa-execution-worker/src/execution.rs [hunk 12]
```diff
@@ -231,7 +231,7 @@ impl ExecutionState {
         // get operation ID
         let operation_id = operation.id;
 
-        // compute fee from (op.max_gas * op.gas_price + op.fee)
+        // compute fee from (op.max_gas + op.fee)
```
Comment (damip) [inline massa-execution-worker/src/execution.rs:234]: total fee is now just `op.fee` and nothing else

File: massa-execution-worker/src/interface_impl.rs [hunk 13]
```diff
@@ -571,7 +571,6 @@ impl Interface for InterfaceImpl {
             validity_start: Slot::new(validity_start.0, validity_start.1),
             validity_end: Slot::new(validity_end.0, validity_end.1),
             max_gas,
-            gas_price: Amount::from_raw(gas_price),
```
Comment (damip) [inline massa-execution-worker/src/interface_impl.rs:574]: * this should be replaced by `fee: Amount`
* massa-sc-runtime should also be updated

File: massa-execution-worker/src/tests/scenarios_mandatories.rs [hunk 14]
```diff
@@ -859,7 +857,6 @@ fn create_execute_sc_operation(
     let op = OperationType::ExecuteSC {
         data: data.to_vec(),
         max_gas: 100_000,
-        gas_price: Amount::from_mantissa_scale(1, 0),
```
Comment (damip) [inline massa-execution-worker/src/tests/scenarios_mandatories.rs:862]: change it to `fee`

File: massa-final-state/src/state_changes.rs [hunk 15]
```diff
@@ -74,7 +74,6 @@ impl Serializer<StateChanges> for StateChangesSerializer {
     ///     destination: Address::from_str("A12htxRWiEm8jDJpJptr6cwEhWNcCSFWstN1MLSa96DDkVM9Y42G").unwrap(),
     ///     handler: String::from("test"),
     ///     max_gas: 10000000,
-    ///     gas_price: Amount::from_str("1").unwrap(),
```
Comment (damip) [inline massa-final-state/src/state_changes.rs:77]: fee

File: massa-final-state/src/state_changes.rs [hunk 16]
```diff
@@ -168,7 +167,6 @@ impl Deserializer<StateChanges> for StateChangesDeserializer {
...
```
Comment (damip) [inline massa-final-state/src/state_changes.rs:171]: fee

File: massa-models/src/operation.rs [hunk 17]
```diff
@@ -448,7 +441,6 @@ impl std::fmt::Display for OperationType {
                 writeln!(f, "\t- target function:{}", target_func)?;
                 writeln!(f, "\t- target parameter:{}", param)?;
                 writeln!(f, "\t- max_gas:{}", max_gas)?;
-                writeln!(f, "\t- gas_price:{}", gas_price)?;
```
Comment (damip) [inline massa-models/src/operation.rs:451]: fee

File: massa-models/src/operation.rs [hunk 18]
```diff
@@ -781,26 +753,11 @@ impl WrappedOperation {
...
-    }
-
-    /// Get the amount of coins used by the operation to pay for gas
-    pub fn get_gas_coins(&self) -> Amount {
-        self.get_gas_price()
-            .saturating_mul_u64(self.get_gas_usage())
-    }
-
     /// Get the total fee paid by the creator
     pub fn get_total_fee(&self) -> Amount {
```
Comment (damip) [inline massa-models/src/operation.rs:757]: this function is now useless. use op.fee directly

File: massa-models/src/operation.rs [hunk 19]
```diff
@@ -831,16 +788,11 @@ impl WrappedOperation {
...
-            } => gas_price.saturating_mul_u64(*max_gas),
-            OperationType::CallSC {
-                max_gas,
-                gas_price,
-                coins,
-                ..
-            } => gas_price
-                .saturating_mul_u64(*max_gas)
+            OperationType::ExecuteSC { max_gas, .. } => {
+                Amount::from_str(&max_gas.to_string()).unwrap_or(Amount::default())
```
Comment (damip) [inline massa-models/src/operation.rs:792]: here we should return just the `coins` property of executesc

also, going through strings is inefficient

File: massa-models/src/operation.rs [hunk 20]
```diff
@@ -831,16 +788,11 @@ impl WrappedOperation {
...
+            }
+            OperationType::CallSC { max_gas, coins, .. } => Amount::from_str(&max_gas.to_string())
```
Comment (damip) [inline massa-models/src/operation.rs:794]: here we should return just the `coins` property of CallSC

also, going through strings is inefficient

**Review by damip — CHANGES_REQUESTED** [review]
There are a few misunderstandings that can be cleared by adding some documentation maybe:
* From now on, only the `fee` parameter represents coins spent as fees, only the `max_gas` parameter represents the amount of gas (computing power) expected to be spent during the execution, and only the size of the operation represents the amount of block space used by the operation
* Coins, gas and block space are three different dimensions (see https://en.wikipedia.org/wiki/Dimensional_analysis )
* The Amount structure encodes Coins and nothing else (already documented for the Amount structure)
* gas is encoded as u64
* block space is encoded as u64
* quotes from https://en.wikipedia.org/wiki/Dimensional_analysis#Dimensional_homogeneity : 
  * One may take ratios of quantities with different dimensions, and multiply or divide them. However, only quantities having the same dimension may be compared, equated, added, or subtracted.
    * this means that `max_gas + fee` is illegal but `max_gas / fee` is legal

**Comment by gterzian** [comment]
@damip Thanks for the clarification! So the `fee` in all cases would come from that found on the operation?

**Comment by damip** [comment]
> @damip Thanks for the clarification! So the `fee` in all cases would come from that found on the operation, except for the `send_message` ABI where we'd have to add it as a call parameter?

Yes exactly

**Comment by damip** [comment]
@gterzian I made a mistake in one of my comments above: the first criterion of the async pool sorting should not be msg.fee but Ratio(msg.fee, max(msg.max_gas,1))

File: massa-models/src/operation.rs (hunk 19 above)
Comment (gterzian) [reply massa-models/src/operation.rs:792]: It seems OperationType::ExecuteSC does not have `coins`, would returning zero be appropriate?

**Comment by gterzian** [comment]
continuing over at https://github.com/massalabs/massa/pull/3173

File: massa-models/src/operation.rs (hunk 19 above)
Comment (damip) [reply massa-models/src/operation.rs:792]: true, let's return Amount::zero() like in the case of RollSell

File: massa-models/src/operation.rs [hunk 21]
```diff
@@ -831,16 +788,11 @@ impl WrappedOperation {
...
+                .unwrap_or(Amount::default())
```
Comment (damip) [inline massa-models/src/operation.rs:795]: clippy warn

File: massa-models/src/operation.rs (hunk 19 above)
Comment (damip) [inline massa-models/src/operation.rs:792]: clippy warn

File: massa-models/src/operation.rs [hunk 22]
```diff
@@ -781,26 +753,11 @@ impl WrappedOperation {
...
-    pub fn get_gas_coins(&self) -> Amount {
-        self.get_gas_price()
-            .saturating_mul_u64(self.get_gas_usage())
-    }
-
     /// Get the total fee paid by the creator
     pub fn get_total_fee(&self) -> Amount {
-        self.get_gas_coins().saturating_add(self.content.fee)
+        self.content.fee.saturating_add(
+            Amount::from_str(&self.get_gas_usage().to_string()).unwrap_or(Amount::default()),
```
Comment (damip) [inline massa-models/src/operation.rs:759]: clippy warn

Comment (damip) [inline massa-models/src/operation.rs:759]: there is a design problem here as well (see other remarks)

File: massa-models/src/operation.rs (hunk 19 above)
Comment (damip) [inline massa-models/src/operation.rs:792]: design problem here

Comment (damip) [reply massa-models/src/operation.rs:792]: > continuing over at #3173

so this PR is abandoned and you continue on 3173 ?

**Comment by gterzian** [comment]
Closing in favor of https://github.com/massalabs/massa/pull/3173

**Pull request closed without merging**